pkg/plugins/tainttoleration/taint_toleration.go

```diff
@@ -0,0 +1,160 @@
+package tainttoleration
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+var _ plugins.Filter = &TaintToleration{}
+
+var TolerationClock = (clock.Clock)(clock.RealClock{})
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
+	description    = "TaintToleration is a plugin that checks if a placement tolerates a managed cluster's taints"
+)
+
+type TaintToleration struct {
+	handle plugins.Handle
+}
+
+func New(handle plugins.Handle) *TaintToleration {
+	return &TaintToleration{
+		handle: handle,
+	}
+}
+
+func (p *TaintToleration) Name() string {
+	return reflect.TypeOf(*p).Name()
+}
+
+func (pl *TaintToleration) Description() string {
+	return description
+}
+
+func (pl *TaintToleration) Filter(ctx context.Context, placement *clusterapiv1beta1.Placement, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, error) {
+	if len(clusters) == 0 {
+		return clusters, nil
+	}
+
+	// do validation on each toleration and return error if necessary
+	for _, toleration := range placement.Spec.Tolerations {
+		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1beta1.TolerationOpExists {
+			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+		}
+		if toleration.Operator == clusterapiv1beta1.TolerationOpExists && len(toleration.Value) > 0 {
+			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+		}
+	}
+
+	existingDecisions := getDecisions(pl.handle, placement)
+
+	// filter the clusters
+	matched := []*clusterapiv1.ManagedCluster{}
+	for _, cluster := range clusters {
```

Comment:
I think we can check if len(taints)==0?

Reply:
I would check if cluster in the decision here.

Reply:
Didn't add len(taints)==0 is because inside `isClusterTolerated`, it will return true once cluster.Spec.Taints == 0. 
Still check if cluster in the decision inside `isTaintTolerated`, as it depends on `taint.Effect == clusterapiv1.TaintEffectNoSelectIfNew`